README.md
## **VCF to HL7 V2 Converter**

### Introduction

VCF-formatted files are the lingua franca of next-generation sequencing, whereas HL7 Version 2 messaging format is the predominant means by which labs send structured results to Electronic Health Records (EHRs). EHRs can directly import HL7 V2 formatted results that conform to the [HL7 Lab Results Interface Implementation Guide](https://www.hl7.org/documentcenter/public/standards/dstu/V251_IG_LRI_R1_STU3_2018JUN.pdf). Here, we provide an open source utility for converting variants from VCF format into HL7 V2 format. Details of the translation logic are on the [manual page](https://github.com/elimuinformatics/vcf2hl7v2/blob/master/docs/Manual.md).

### Install
Before installing vcf2hl7v2 you need to install cython and wheel.
```
pip install cython wheel
```
Now, install vcf2hl7v2 binary from pip.
```
pip install --index-url https://test.pypi.org/simple/ --extra-index-url https://pypi.org/simple/ vcf2hl7v2
```

### Quick Examples
(some sample VCF files are [here](https://github.com/elimuinformatics/vcf2hl7v2/tree/master/vcf2hl7v2/test))

```python
>>> import vcf2hl7v2
>>> vcf_hl7v2_converter = vcf2hl7v2.Converter('sample.vcf', 'GRCh37')
>>> vcf_hl7v2_converter.convert()
```

### Logging

You can use python standard [logging](https://docs.python.org/3/library/logging.html) to enable logs. Two loggers ('vcf2hl7v2.general') and ('vcf2hl7v2.invalidrecord') are available to configure.
* **vcf2hl7v2.general**: standard library logs. 
* **vcf2hl7v2.invalidrecord**: logs all the records from vcf file which are in conversion region but are not converted to HL7 V2 format.

```python
>>> import logging
# create logger
>>> logger = logging.getLogger('vcf2hl7v2.invalidrecord')
>>> logger.setLevel(logging.DEBUG)
# create console handler and set level to debug
>>> ch = logging.FileHandler('invalidrecord.log')
>>> ch.setLevel(logging.DEBUG)
# create formatter
>>> formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')
# add formatter to ch
>>> ch.setFormatter(formatter)
# add ch to logger
>>> logger.addHandler(ch)
```

### Scope

Software converts somatic and germline simple variants (SNVs, MNVs, Indels), along with zygosity and phase relationships, for autosomes, sex chromosomes, and mitochondrial DNA. Select clinical annotations (e.g. clinical significance, phenotype), supplied in a separate annotation file are incorporated into the conversion.

* Not supported
    * **Structural variants**: Software does not support conversion of structural variants (where INFO.SVTYPE is present). 
    * **Alt contigs**: Software does not support conversion of variants aligned to Alt contigs. We recommend caution in using this software against VCFs generated with an alternate-locus aware variant caller, as variants mapped to Alt contigs will not be converted.
    * **Query liftover**: Software assumes that regions (conversion region, studied region) and VCF are based on the same genomic build. 
    * **Chromosome synonyms (e.g. '1' vs. 'chr1')**: Software assumes that chromosome representation is consistent between regions (e.g. in BED files) and VCF. For instance, if VCF uses 'chr1', then BED file must also use 'chr1' 

### License and Limitations

Software is available for use under an [Apache 2.0 license](https://opensource.org/licenses/Apache-2.0), and is intended solely for experimental use, to help further Genomics-EHR integration exploration. Software is expressly not ready to be used with identifiable patient data or in delivering care to patients. Code issues should be tracked here. Comments and questions can also be directed to info@elimu.io or srikarchamala@gmail.com.


CHANGELOG.md
# v0.0.1 (TBD)

Converts VCF variants into HL7 V2 format that conform to the [HL7 Lab Results Interface Implementation Guide](https://www.hl7.org/documentcenter/public/standards/dstu/V251_IG_LRI_R1_STU3_2018JUN.pdf). 

## Added
* In scope are somatic and germline simple variants (SNVs, MNVs, Indels), along with zygosity and phase relationships, for autosomes, sex chromosomes, and mitochondrial DNA.
* Supports VCF file (text-based or bgzipped) and optionally tabix files for query.
* Supports genome build ('GRCh37' or 'GRCh38');
* Optionally supports a query region in the form of .bed or dictionary that indicates the region(s) to convert.
* Optionally supports inclusion of  'region-studied' observations that detail which portions of the conversion region were studied.
* Optionally supports clinical annotations (e.g. clinical significance, phenotype), supplied in a separate annotation file.


vcf2hl7v2/test/test_unit.py
import doctest
import unittest
import vcf2hl7v2
from vcf2hl7v2.common import *

suite = doctest.DocTestSuite(vcf2hl7v2)


class TestChromIdentifier(unittest.TestCase):

    def test_chrom_1_22(self):
        actual_chrom = ['chr1', '1', 'CHR1', '22', 'CHR22', 'chr22']
        expected_chrom = ['1', '1', '1', '22', '22', '22']
        i = 0
        for chrom in actual_chrom:
            self.assertEqual(extract_chrom_identifier(
                chrom), expected_chrom[i])
            i += 1

    def test_chrom_X_Y(self):
        actual_chrom = ['chrX', 'X', 'x', 'CHRX', 'chrY', 'Y', 'Y', 'CHRY']
        expected_chrom = ['X', 'X', 'X', 'X', 'Y', 'Y', 'Y', 'Y']
        i = 0
        for chrom in actual_chrom:
            self.assertEqual(extract_chrom_identifier(
                chrom), expected_chrom[i])
            i += 1

    def test_chrom_M(self):
        actual_chrom = ['MT', 'm', 'M', 'mt', 'chrm', 'chrM', 'chrmt', 'chrMT']
        expected_chrom = ['M', 'M', 'M', 'M', 'M', 'M', 'M', 'M']
        i = 0
        for chrom in actual_chrom:
            self.assertEqual(extract_chrom_identifier(
                chrom), expected_chrom[i])
            i += 1

    def test_chrom_validation(self):
        actual_chrom = [
            'chrX', 'R', 'mt', 'chrR', 'chrM', 'chrMT',
            'chr30', 'P', 'Z', '45', 'o', 'CHRX', '1', '22', 'CHR21', 'x']
        recognized = [
            True, False, True, False, True, True,
            False, False, False, False, False, True, True, True, True, True]
        i = 0
        for chrom in actual_chrom:
            self.assertEqual(validate_chrom_identifier(
                chrom), recognized[i])
            i += 1


class TestDataType(unittest.TestCase):

    def test_validate_ratio_ad_dp(self):
        ratio_ad_dp = [0.2, 0.65, 0.99, 1.3, -0.7, "Chr", False, True]
        valid = [True, True, True, False, False, False, False, False]
        i = 0
        for value in ratio_ad_dp:
            self.assertEqual(validate_ratio_ad_dp(
                value), valid[i])
            i += 1

    def test_validate_has_tabix(self):
        has_tabix = [True, False, 1, 23.4, 'C', "CHROM"]
        valid = [True, True, False, False, False, False]
        i = 0
        for value in has_tabix:
            self.assertEqual(validate_has_tabix(
                value), valid[i])
            i += 1

    def test_validate_seed(self):
        seed = [1, 2000, 'seed', True, 4000, 1234]
        valid = [True, True, False, False, True, True]
        i = 0
        for value in seed:
            self.assertEqual(validate_seed(
                value), valid[i])
            i += 1

    def test_get_alphabet_index(self):
        index = [1, 26, 29, 79, 207, 703, 1234]
        unique_identifier = ['a', 'z', 'ac', 'ca', 'gy', 'aaa', 'aul']
        i = 0
        for value in index:
            self.assertEqual(get_alphabet_index(
                value), unique_identifier[i])
            i += 1


suite.addTests(unittest.TestLoader(
).loadTestsFromTestCase(TestChromIdentifier))
suite.addTests(unittest.TestLoader(
).loadTestsFromTestCase(TestDataType))

if __name__ == '__main__':
    unittest.main()


vcf2hl7v2/test/test_integration.py
import doctest
import unittest
import os
import vcf2hl7v2
from os.path import join, dirname
import shutil
import logging
import filecmp

suite = doctest.DocTestSuite(vcf2hl7v2)


class TestVcf2HL7v2Inputs(unittest.TestCase):

    def test_required_vcf_filename(self):
        with self.assertRaises(Exception) as context:
            vcf2hl7v2.Converter()
        self.assertTrue(
            'You must provide vcf_filename' in str(context.exception))

    def test_invalid_vcf_filename(self):
        self.assertRaises(FileNotFoundError,
                          vcf2hl7v2.Converter, *['Hello', 'GRCh38'])

    def test_required_ref_build(self):
        with self.assertRaises(Exception) as context:
            vcf2hl7v2.Converter(os.path.join(
                os.path.dirname(__file__), 'vcf_example1.vcf'))
        self.assertEqual(
            'You must provide build number ("GRCh37" or "GRCh38")', str(
                context.exception))

    def test_invalid_ref_build(self):
        with self.assertRaises(Exception) as context:
            vcf2hl7v2.Converter(os.path.join(
                os.path.dirname(__file__), 'vcf_example1.vcf'), 'b38')
        self.assertEqual(
            'You must provide build number ("GRCh37" or "GRCh38")', str(
                context.exception))

    def test_valid_ref_build_37(self):
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(os.path.join(
            os.path.dirname(__file__), 'vcf_example1.vcf'), 'GRCh37')
        self.assertEqual(type(o_vcf_2_hl7v2), vcf2hl7v2.Converter)

    def test_valid_ref_build_38(self):
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(os.path.join(
            os.path.dirname(__file__), 'vcf_example1.vcf'), 'GRCh38')
        self.assertEqual(type(o_vcf_2_hl7v2), vcf2hl7v2.Converter)

    def test_conv_region_only(self):
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_example3.bed')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh37',
            'abc',
            conv_region_filename=conv_region_filename)
        self.assertEqual(type(o_vcf_2_hl7v2), vcf2hl7v2.Converter)

    def test_conv_region_dict(self):
        conv_region_dict = {
            "Chromosome": ["X", "X", "M"],
            "Start": [50000, 55000, 50000],
            "End": [52000, 60600, 60025]
        }
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh37',
            'abc',
            conv_region_dict=conv_region_dict)
        self.assertEqual(type(o_vcf_2_hl7v2), vcf2hl7v2.Converter)

    def test_conv_region_region_studied(self):
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example3.bed')
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_example3.bed')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh37',
            'abc',
            conv_region_filename=conv_region_filename,
            region_studied_filename=region_studied_filename)
        self.assertEqual(type(o_vcf_2_hl7v2), vcf2hl7v2.Converter)

    def test_no_conv_region_region_studied(self):
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example3.bed')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh37',
            'abc',
            region_studied_filename=region_studied_filename)
        self.assertEqual(type(o_vcf_2_hl7v2), vcf2hl7v2.Converter)


class TestTranslation(unittest.TestCase):
    @classmethod
    def setUpClass(self):
        self.TEST_RESULT_DIR = os.path.join(
            os.path.dirname(__file__), 'output')
        if os.path.exists(self.TEST_RESULT_DIR):
            shutil.rmtree(self.TEST_RESULT_DIR)
        os.mkdir(self.TEST_RESULT_DIR)

    @classmethod
    def tearDownClass(self):
        shutil.rmtree(self.TEST_RESULT_DIR)

    def test_wo_patient_id(self):
        self.maxDiff = None
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(os.path.join(
            os.path.dirname(__file__), 'vcf_example1.vcf'), 'GRCh37')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_wo_patient_example1.txt')
        expected_output_filename = os.path.join(
            os.path.dirname(__file__), 'expected_example1_wo_patient.txt')
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )

    def test_with_patient_id(self):
        self.maxDiff = None
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(os.path.join(os.path.dirname(
            __file__), 'vcf_example1.vcf'), 'GRCh37', 'HG00628')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_with_patient_example1.txt')
        expected_output_filename = os.path.join(os.path.dirname(
            __file__), 'expected_example1_with_patient.txt')
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )

    def test_region_studied(self):
        self.maxDiff = None
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example3.bed')
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_example3.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_example3.txt')
        expected_output_filename = os.path.join(
            os.path.dirname(__file__), 'expected_example3.txt')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh38',
            'HG00628',
            conv_region_filename=conv_region_filename,
            region_studied_filename=region_studied_filename)
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )

    def test_region_studied_dict(self):
        conv_region_dict = {
            "Chromosome": ["X", "X", "M"],
            "Start": [50000, 55000, 50000],
            "End": [52000, 60600, 60025]
        }
        self.maxDiff = None
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example3.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_example3_dict.txt')
        expected_output_filename = os.path.join(
            os.path.dirname(__file__), 'expected_example3.txt')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh38',
            'HG00628',
            conv_region_dict=conv_region_dict,
            region_studied_filename=region_studied_filename)
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )

    # Check if region studied observation outside the vcf files are also
    # included in hl7v2 report.
    def test_multiple_region_studied(self):
        self.maxDiff = None
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example4.bed')
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_example4.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_example4.txt')
        expected_output_filename = os.path.join(
            os.path.dirname(__file__), 'expected_example4.txt')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example4.vcf'),
            'GRCh38',
            'HG00628',
            conv_region_filename=conv_region_filename,
            region_studied_filename=region_studied_filename)
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )

    def test_region_studied_only(self):
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example4.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_example4_test.txt')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example4.vcf'),
            'GRCh38',
            'HG00628',
            region_studied_filename=region_studied_filename)
        o_vcf_2_hl7v2.convert(output_filename)

    def test_empty_hl7v2_txt(self):
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_empty_example4.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_example4_test.txt')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example4.vcf'),
            'GRCh38',
            'HG00628',
            conv_region_filename=conv_region_filename)
        o_vcf_2_hl7v2.convert(output_filename)

    def test_tabix(self):
        self.maxDiff = None
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example4.bed')
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_example4.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'hl7v2_example4_tabix.txt')
        expected_output_filename = os.path.join(
            os.path.dirname(__file__), 'expected_example4.txt')
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example4.vcf.gz'),
            'GRCh38',
            'HG00628',
            has_tabix=True,
            conv_region_filename=conv_region_filename,
            region_studied_filename=region_studied_filename)
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )

    def test_annotation(self):
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
                vcf_filename=os.path.join(
                                        os.path.dirname(__file__),
                                        'NB6TK328_filtered.vcf'
                                    ),
                ref_build='GRCh38',
                patient_id='NB6TK328',
                has_tabix=False,
                conv_region_filename=os.path.join(
                                        os.path.dirname(__file__),
                                        'NB6TK328_conversion_region.bed'
                                    ),
                conv_region_dict=None,
                region_studied_filename=os.path.join(
                                        os.path.dirname(__file__),
                                        'NB6TK328_region_studied.bed'
                                    ),
                ratio_ad_dp=0.95,
                source_class='germline',
                seed=5000,
                annotation_filename=os.path.join(
                                        os.path.dirname(__file__),
                                        'NB6TK328_annotations.txt'
                                    ),
            )
        output_filename = os.path.join(os.path.dirname(
            __file__), self.TEST_RESULT_DIR, 'NB6TK328.txt')
        expected_output_filename = os.path.join(
            os.path.dirname(__file__), 'expected_NB6TK328.txt')
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(
                filecmp.cmp(output_filename, expected_output_filename),
                True
            )


class TestLogger(unittest.TestCase):
    def setUp(self):
        # create file handler and set level to debug
        self.log_general_filename = os.path.join(
            os.path.dirname(__file__), self.LOG_DIR, 'general.log')
        self.log_invalid_record_filename = os.path.join(
            os.path.dirname(__file__), self.LOG_DIR, 'invalid_record.log')
        genearl_fh = logging.FileHandler(self.log_general_filename)
        invalid_record_fh = logging.FileHandler(
            self.log_invalid_record_filename)
        genearl_fh.setLevel(logging.DEBUG)
        invalid_record_fh.setLevel(logging.DEBUG)
        # create formatter
        formatter = logging.Formatter(
            '%(asctime)s - %(name)s - %(levelname)s - %(message)s')
        # add formatter to fh
        genearl_fh.setFormatter(formatter)
        invalid_record_fh.setFormatter(formatter)
        # store as a class variable
        self.genearl_fh = genearl_fh
        self.invalid_record_fh = invalid_record_fh

    @classmethod
    def setUpClass(self):
        self.LOG_DIR = os.path.join(os.path.dirname(__file__), 'log')
        if os.path.exists(self.LOG_DIR):
            shutil.rmtree(self.LOG_DIR)
        os.mkdir(self.LOG_DIR)

    # TODO: Delete the log folder after running
    # all the tests, below method throws error becasue
    # log files are still in use.
    # @classmethod
    # def tearDownClass(self):
    #     shutil.rmtree(self.LOG_DIR)

    def test_logger_forks(self):
        region_studied_filename = os.path.join(
            os.path.dirname(__file__), 'RegionsStudied_example3.bed')
        conv_region_filename = os.path.join(os.path.dirname(
            __file__), 'RegionsToConvert_example3.bed')
        output_filename = os.path.join(os.path.dirname(
            __file__), self.LOG_DIR, 'logging_hl7v2.txt')
        # create logger
        general_logger = logging.getLogger('vcf2hl7v2.general')
        invalid_record_logger = logging.getLogger('vcf2hl7v2.invalidrecord')
        general_logger.setLevel(logging.DEBUG)
        invalid_record_logger.setLevel(logging.DEBUG)
        # add ch to logger
        general_logger.addHandler(self.genearl_fh)
        invalid_record_logger.addHandler(self.invalid_record_fh)
        o_vcf_2_hl7v2 = vcf2hl7v2.Converter(
            os.path.join(
                os.path.dirname(__file__),
                'vcf_example3.vcf'),
            'GRCh38',
            'HG00628',
            conv_region_filename=conv_region_filename,
            region_studied_filename=region_studied_filename)
        o_vcf_2_hl7v2.convert(output_filename)
        self.assertEqual(os.path.exists(self.log_general_filename), True)
        self.assertEqual(os.path.exists(
            self.log_invalid_record_filename), True)


suite.addTests(
    unittest.TestLoader().loadTestsFromTestCase(TestVcf2HL7v2Inputs))
suite.addTests(unittest.TestLoader().loadTestsFromTestCase(TestTranslation))
suite.addTests(unittest.TestLoader().loadTestsFromTestCase(TestLogger))


if __name__ == '__main__':
    unittest.main()


vcf2hl7v2/converter.py
import vcf
import pyranges
from .hl7v2_message_generator import _get_hl7v2_message
import logging
import sys
import pandas as pd
from .common import *
general_logger = logging.getLogger('vcf2hl7v2.general')
"""Converter for a VCF version >4.1 file."""


class Converter(object):
    """Creates a new Converter Object to convert a VCF file.

    Parameters
    ----------

    **vcf_filename** (required): Path to a text-based or bgzipped VCF
    file. Valid path and filename without whitespace must be provided.
    VCF file must conform to VCF Version 4.1 or later. FORMAT.GT must be
    present. Multi-sample VCFs are allowed, but only the first sample
    will be converted.
    has_tabix (required if VCF file is bgzipped): Set to 'True' if
    there is a tabix index. Tabix file must have the same name as the
    bgzipped VCF file, with a '.tbi' extension, and must be in the same
    folder.

    **ref_build** (required): Genome Reference Consortium genome
    assembly to which variants in the VCF were called. Must be one of
    'GRCh37' or 'GRCh38'.

    **patient_id** (optional): Supplied patient ID is inserted into
    generated HL7V2 output. Alphanumeric string without whitespace. if \
    not provided,
    header of first sample column is used.

    **Conversion region** (optional): Subset of the VCF file to be
    converted into HL7V2. If absent, the entire VCF file is converted. Can
    be supplied as either a parameter (conv_region_dict) or as a BED
    file (conv_region_filename):

       **conv_region_dict** : Array of regions (e.g. '{"Chromosome":
       ["X", "X", "M"],"Start": [50000, 55000, 50000],"End": [52000,
       60600, 60025]}'). Values for Chromosome must align with values in
       VCF #CHROM field. Ranges must be `0-based \
       <https://www.biostars.org/p/84686/>`_
       (or 0-start, half-open) and based on GRCh37 or GRCh38 \
       reference sequences.

       **conv_region_filename**: Valid path and filename without
       whitespace must be provided. Must be a valid BED file with first 3
       columns: <chr> <start> <stop>. Values in <chr> field must align
       with values in VCF #CHROM field. Ranges must be based on GRCh37 or
       GRCh38 reference sequences. Note that BED files are
       `0-based <https://www.biostars.org/p/84686/>`_ (or 0-start,
       half-open) whereas VCF files and HL7V2 output are 1-based (or
       1-start, fully-closed).

    **annotation_filename** (optional): Valid path and filename without
    whitespace must be provided. Must be a valid tab-delimited text file.
    Columns 1-4 are CHROM, POS, REF, ALT and must match a row in the
    VCF. Columns 5-11 are gene, transcriptRefSeq, cHGVS, proteinRefSeq,
    pHGVS, clinSig, phenotype. All columns must be present. Columns 1-4
    must be populated. Columns 5-11 can contain nulls.

    **region_studied_filename** (optional): Subset of patient's genome
    that was studied in the generation of the VCF file. Valid path and
    filename without whitespace must be provided. Must be a valid BED
    file, with first 3 columns: <chr> <start> <stop>. Values in <chr>
    field must align with values in VCF #CHROM field. Ranges must be
    based on GRCh37 or GRCh38 reference sequences. Note that BED files
    are `0-based <https://www.biostars.org/p/84686/>`_ (or 0-start,
    half-open) whereas VCF files and HL7V2 output are 1-based (or 1-start,
    fully-closed).

    **ratio_ad_dp** (optional)(default value = 0.99): This ratio \
    determine whether to assign Homoplasmic or Heteroplasmic.
    If allelic depth (FORMAT.AD) / read depth (FORMAT.DP) is \
    greater than ratio_ad_dp then allelic state is
    homoplasmic; else heteroplasmic.

    **source_class** (optional)(default value = germline): An \
    assertion as to whether variants in the VCF file are in the \
    germline (i.e. inherited), somatic (e.g. arose spontaneously \
    as cancer mutations), or mixed (i.e. may be a combination of \
    germline and/or somatic).

    **seed** (optional)(default value = 1000): Used to set the \
    starting integer count for OBX-1 (sequence number)
    Returns
    -------

    Object
    An Instance of Converter that helps to convert vcf file.

    """

    def __init__(
            self, vcf_filename=None, ref_build=None, patient_id=None,
            has_tabix=False, conv_region_filename=None, conv_region_dict=None,
            annotation_filename=None, region_studied_filename=None,
            ratio_ad_dp=0.99, source_class='germline', seed=1000):

        super(Converter, self).__init__()
        if not (vcf_filename):
            raise Exception('You must provide vcf_filename')
        if not ref_build or ref_build not in ["GRCh37", "GRCh38"]:
            raise Exception(
                'You must provide build number ("GRCh37" or "GRCh38")')
        self.vcf_filename = vcf_filename
        try:
            self._vcf_reader = vcf.Reader(filename=vcf_filename)
        except FileNotFoundError:
            raise
        except BaseException:
            self._generate_exception("Please provide valid 'vcf_filename'")
        if not patient_id:
            patient_id = self._vcf_reader.samples[0]
        if conv_region_filename:
            try:
                self.conversion_region = pyranges.read_bed(
                    conv_region_filename)
            except FileNotFoundError:
                raise
            except BaseException:
                self._generate_exception(
                    "Please provide valid 'conv_region_filename'")
        elif conv_region_dict:
            try:
                self.conversion_region = pyranges.from_dict(conv_region_dict)
            except FileNotFoundError:
                raise
            except BaseException:
                self._generate_exception(
                    "Please provide valid 'conv_region_dict'")
        else:
            self.conversion_region = None
        self.annotation_filename = annotation_filename
        if self.annotation_filename is None:
            self.annotations = None
        else:
            try:
                self.annotations = pd.read_csv(
                            self.annotation_filename,
                            names=[
                                'CHROM', 'POS', 'REF', 'ALT', 'gene',
                                'transcriptRefSeq', 'cHGVS', 'proteinRefSeq',
                                'pHGVS', 'clinSig', 'phenotype'
                            ],
                            sep='\t'
                        )
            except FileNotFoundError:
                raise
            except BaseException:
                self._generate_exception(
                        "Please provide valid 'annotation_filename'"
                    )
        if region_studied_filename:
            try:
                self.region_studied = pyranges.read_bed(
                    region_studied_filename)
            except FileNotFoundError:
                raise
            except BaseException:
                self._generate_exception(
                    "Please provide valid 'region_studied_filename'")
        else:
            self.region_studied = None

        if not validate_has_tabix(has_tabix):
            raise Exception("Please provide a valid 'has_tabix'")

        if not validate_ratio_ad_dp(ratio_ad_dp):
            raise Exception("Please provide a valid 'ratio_ad_dp'")

        if not validate_seed(seed):
            raise Exception("Please provide a valid seed")

        if source_class.title() not in Genomic_Source_Class.set_():
            raise Exception(
                'Please provide a valid Source Class ("germline" or "somatic")'
            )

        self.ratio_ad_dp = ratio_ad_dp
        self.has_tabix = has_tabix
        self.patient_id = patient_id
        self.ref_build = ref_build
        self.conv_region_filename = conv_region_filename
        self.source_class = source_class
        self.seed = seed
        general_logger.info("Converter class instantiated successfully")

    def convert(self, output_filename='hl7v2.txt'):
        """ Generates HL7 V2 format data as output_filename \
        or hl7v2.txt if it is not provided

        Parameters
        ----------
        output_filename:
            Path to output hl7v2 text file.

        """
        general_logger.info("Starting VCF to HL7V2 Conversion")
        _get_hl7v2_message(
            list(self._vcf_reader), self._vcf_reader, self.ref_build,
            self.patient_id, self.has_tabix, self.conversion_region,
            self.source_class, self.region_studied, self.ratio_ad_dp,
            self.annotations, self.seed, output_filename)
        general_logger.info("Completed VCF to HL7V2 Conversion")

    def _generate_exception(self, msg):
        general_logger.error(msg, exc_info=True)
        raise Exception(msg, sys.exc_info)
